pep-9999.rst

```diff
@@ -0,0 +1,210 @@
+PEP: 9999
+Title: Remove wstr from Unicode
+Author: Inada Naoki <songofacandy@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 24-Jun-2020
+Python-Version: 3.10
+
+
+Abstract
+========
+
+PEP 393 deprecated some unicode APIs, and introduced ``wchar_t *wstr``,
+and ``Py_ssize_t wstr_length`` in unicode implementation for backward
+compatibility of these deprecated APIs. [1]_
+
+This PEP is planning removal of ``wstr``, and ``wstr_length`` with
+deprecated APIs using these members.
+
+
+Motivation
+==========
+
+Memory usage
+------------
+
+``str`` is one of the most used types in Python.  Even most simple ASCII
+strings have a ``wstr`` member.  It consumes 8 bytes on 64bit systems.
+
+
+Runtime overhead
+----------------
+
+To support legacy Unicode object created by
+``PyUnicode_FromUnicode(NULL, length)``, many Unicode APIs has
+``PyUnicode_READY()`` check.
+
+When we drop support of legacy unicode object, We can reduce this overhead
+too.
+
+
+Simplicity
+----------
+
+Support of legacy Unicode object makes Unicode implementation complex.
+Until we drop legacy Unicode object, it is very hard to try other Unicode
+implementation like UTF-8 based implementation in PyPy.
+
+
+Plan
+=====
+
+Python 3.9
+----------
+
+These macros and functions are marked as deprecated, using
+``Py_DEPRECATED`` macro.
+
+* Py_UNICODE_WSTR_LENGTH
+* PyUnicode_GET_SIZE
+* PyUnicode_GET_DATA_SIZE
+* PyUnicode_AS_UNICODE
+* PyUnicode_AS_DATA
+* PyUnicode_AsUnicode
+* _PyUnicode_AsUnicode
+* PyUnicode_AsUnicodeAndSize
+* PyUnicode_FromUnicode
+
+
+Python 3.10
+-----------
+
+Behavior changed:
+
+* ``PyUnicode_FromUnicode(NULL, size)`` and
+  ``PyUnicode_FromStringAndSize(NULL, size))`` will raise
+  DeprecationWarning when ``size > 0``.
+
+* PyArg_ParseTuple and PyArg_ParseTupleAndKeywords will raise
+  DeprecationWarning when 'u', 'u#', 'Z', and 'Z#' format are used.
+
+
+Python 3.12
+-----------
+
+Following members will be removed from Unicode strucutre:
+
+* ``wstr``
+* ``wstr_length``
+* ``state.compact``
+* ``state.ready``
+
+Following macros and functions will be removed:
+
+* Py_UNICODE_WSTR_LENGTH
+* PyUnicode_GET_SIZE
+* PyUnicode_GET_DATA_SIZE
+* PyUnicode_AS_UNICODE
+* PyUnicode_AS_DATA
+* PyUnicode_AsUnicode
+* _PyUnicode_AsUnicode
+* PyUnicode_AsUnicodeAndSize
+* PyUnicode_FromUnicode
+
+Behavior changed:
+
+* ``PyUnicode_FromStringAndSize(NULL, size))`` will raise
+  RuntimeError when ``size > 0``.
+
+* PyArg_ParseTuple and PyArg_ParseTupleAndKeywords will raise
+  RuntimeError when 'u', 'u#', 'Z', and 'Z#' formats are used.
+
+Following macros, enum members will be marked deprecated.
+
+* PyUnicode_WCHAR_KIND
```

Comment:
Is it possible to mark an enum as deprecated with Py_DEPRECATED()? Is it just about marking it as deprecated in the documentation?

Reply:
I am not sure about it.  If we can not use Py_DEPRECATED, we will use `/* deprecated */` comment and document.